doozerlib/operator_metadata.py

```diff
@@ -264,20 +264,63 @@ def replace_version_by_sha_on_image_references(self, arch):
             with open(file, 'r') as reader:
                 contents = reader.read()
 
-            new_contents = re.sub(
-                r'{}/([^:]+):([^\'"\s]+)'.format(self.operator_csv_registry),
-                lambda i: '{}/{}@{}'.format(
-                    self.operator_csv_registry,
-                    i.group(1),
-                    self.fetch_image_sha('{}:{}'.format(i.group(1), i.group(2)), arch)
-                ),
-                contents,
-                flags=re.MULTILINE
-            )
+            new_contents = self.find_and_replace_image_versions_by_sha(contents, arch)
 
             with open(file, 'w') as writer:
                 writer.write(new_contents)
 
+    @log
+    def find_and_replace_image_versions_by_sha(self, contents, arch):
+        """Read "contents" collecting all image references, query the corresponding
+        SHA for each found image and replace them inline.
+
+        :param contents: a string with the contents of a YAML file that might have image references
+        :param arch: string with an architecture or "manifests-list", used when picking SHAs
+        :return: contents string back, with image references replaced + "relatedImages" node under "spec"
+        """
+        found_images = {}
+
+        def collect_replaced_image(match):
+            image = '{}/{}@{}'.format(
+                self.operator_csv_registry,
+                match.group(1),
+                self.fetch_image_sha('{}:{}'.format(match.group(1), match.group(2)), arch)
+            )
+            key = u'{}'.format(re.search(r'([^/]+)/(.+)', match.group(1)).group(2))
+            found_images[key] = u'{}'.format(image)
+            return image
+
+        new_contents = re.sub(
+            r'{}/([^:]+):([^\'"\s]+)'.format(self.operator_csv_registry),
+            collect_replaced_image,
+            contents,
+            flags=re.MULTILINE
+        )
+
+        new_contents = self.append_related_images_spec(new_contents, found_images)
+        return new_contents
+
+    @log
+    def append_related_images_spec(self, contents, images):
+        """Create a new node inside "spec" listing all related images, without
+        parsing the YAML, to avoid unwanted modifications when re-serializing it.
+
+        :param contents: CSV YAML string
+        :param images: a dict containing images (key: name, value: image)
+        :return: contents string back with "relatedImages" node under "spec"
+        """
+        related_images = []
+        for name, image in images.items():
+            related_images.append('    {}: {}'.format(name, image))
+        related_images.sort()
+
+        return re.sub(
```

Comment:
This seems error-prone. Can we use a yaml parser to manipulate the YAML file instead of text-base processing?

Reply:
Hi @vfreex , parsing the YAML would make my life **much** easier ... unfortunately `parsing -> modifying -> serializing to file again` produces undesired modifications on the CSV, specially in multiline texts, such as `description:`.
That's why I'm modifying it "the hard way".

Reply:
@thiagoalessio Using a YAML parser is not required but nice to have.
Serialized YAML should be equivalent to the original one, even if its layout may change.
There is another YAML parser [ruamel.yaml](https://yaml.readthedocs.io/en/latest/), which can supports round-trip preservation of comments and layout. If you have time to try, it might be a good investigation.

Reply:
nice stuff @vfreex ! thanks for sharing. I'll try it out in a copy of this branch, but I guess it would already be beneficial to have this PR merged, so we can make @shawn-hurley happy :tada: